### DISCUSSION TITLE
Performance changes over time

### DISCUSSION THREAD

**Discussion by excubo-jg** [microsoft/pyright#5651] [General]
I have noticed a significant drop in responsiveness in the IDE with v316 with processing times being 50%-200% higher. Out of curiosity I did run a benchmark of pyright version's processing speed on two repos of mine as-is. There have been several instances (293, 312, 314) when both repos took a performance hit, v317 bringing one of them back to v27x levels. I was surprised by the magnitude of the changes (especially the 2nd repo is quite jumpy). As this is more of a snapshot I am wondering if speed is tracked on a more significant set of repos and is this a general trend? It seems to me that the changes in CLI performance are impacted by the time needed for the libraries used. https://github.com/microsoft/pyright/discussions/5496 points in that direction and also shows the improvement with v317 (and the deterioration with 312 and 314).

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/Admin/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/Admin/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">

version | sec | sec | % | %
-- | -- | -- | -- | --
320 | 18.2 | 9.0 | 202 | 110
318 | 18.1 | 8.8 | 202 | 107
317 | 18.0 | 8.9 | 201 | 108
316 | 27.0 | 23.7 | 300 | 289
315 | 27.3 | 24.0 | 304 | 292
314 | 23.7 | 26.1 | 264 | 318
313 | 17.2 | 16.7 | 191 | 204
312 | 17.1 | 16.8 | 190 | 204
311 | 14.5 | 13.3 | 161 | 162
310 | 14.7 | 13.2 | 164 | 160
309 | 15.9 | 13.0 | 177 | 158
308 | 15.8 | 13.2 | 176 | 161
307 | 15.7 | 12.3 | 174 | 150
306 | 14.9 | 13.1 | 165 | 159
305 | 15.2 | 9.2 | 169 | 112
304 | 15.3 | 9.3 | 170 | 113
303 | 15.1 | 9.4 | 168 | 114
302 | 15.5 | 13.3 | 173 | 161
301 | 15.8 | 13.0 | 175 | 158
300 | 15.5 | 12.0 | 173 | 146
299 | 15.4 | 11.8 | 171 | 143
298 | 15.3 | 11.9 | 170 | 145
297 | 15.4 | 11.5 | 171 | 140
296 | 15.3 | 11.6 | 170 | 141
295 | 15.4 | 11.5 | 172 | 139
294 | 15.2 | 11.5 | 169 | 140
293 | 14.3 | 11.3 | 159 | 138
292 | 11.6 | 8.2 | 129 | 100
291 | 10.2 | 7.4 | 114 | 90
290 | 10.5 | 11.0 | 116 | 134
289 | 10.5 | 11.1 | 117 | 135
288 | 10.4 | 10.2 | 116 | 124
287 | 10.3 | 28.6 | 114 | 347
286 | 9.5 | 10.1 | 106 | 123
285 | 9.4 | 8.2 | 105 | 100
284 | 9.5 | 9.6 | 105 | 116
283 | 9.4 | 8.8 | 104 | 107
282 | 9.6 | 22.5 | 106 | 273
281 | crash | 7.6 | crash | 92
280 | 9.3 | 7.4 | 103 | 90
279 | 9.5 | 7.9 | 105 | 96
278 | 9.4 | 7.5 | 105 | 92
277 | 9.2 | 8.8 | 102 | 107
276 | 9.1 | 8.2 | 102 | 99
275 | 9.2 | 8.4 | 102 | 102
274 | 9.1 | 7.8 | 101 | 95
273 | 9.0 | 8.7 | 100 | 105
272 | 8.9 | 8.2 | 99 | 99
271 | 9.0 | 8.2 | 100 | 100

</body>

</html>

**Comment by erictraut** [comment]
Thanks for providing the data. That's really interesting.

Pyright is designed with performance in mind, and I take performance regressions very seriously. I track type checking performance closely for a number of repos. The pylance team also tracks the performance of language server features (completion suggestions, etc.) on a set of representative code bases. These benchmarks run on every commit to both pyright and pylance, so we're able to attribute performance regressions to a specific commit. These benchmarks also track other metrics like peak memory usage and cold start times.

The analysis times you've measured across pyright versions vary way more than any of the benchmarks we run. I wonder if you're using coding patterns or types that are not well represented in other benchmark or test suites. If so, it would be good to include one or more of your repos. Are your repos public? If so, I'd encourage you to add them to the [mypy_primer projects list](https://github.com/hauntsaninja/mypy_primer/blob/master/mypy_primer/projects.py). Mypy_primer is a tool created by @hauntsaninja, and it's integrated into pyright's CI pipeline (as well as mypy and typeshed). It is designed mostly to catch behavioral regressions, but it can also catch egregious performance regressions.

Looking at the specific releases that appear to be outliers in your list...

v293: There are two changes that stand out as having potential for performance impacts: 1) an enhancement where I added support for overloaded functions captured by a ParamSpec, and 2) an update to the typeshed stubs.

v312: There are two changes that stand out as having potential for performance impacts: 1) a bug fix relating to calls that involve the same generic function multiple times, and 2) an update to the typeshed stubs.

v314: There are two changes that stand out as having potential for performance impacts: 1) a bug fix that affects protocol matching, and 2) an update to the typeshed stubs.

I wouldn't expect any of the changes above to have an appreciable effect on performance, so I'm surprised by the results you've measured.

I'm interested in your methodology. Did you gather all of your measurements on the same computer running on "bare metal", or was it done in the cloud on a VM? Were there any other background processes running at the time you collected these numbers? Did you use the same version of node for all runs? Did you use the same Python environment populated with the same versions of libraries for each of them? All of these factors can affect the analysis timing and could explain the significant variability you're seeing if you are not accounting for these factors.

When I'm focusing on performance analysis, I typically run each test three times, and I throw out the first two runs because they tend to measure disk and file system performance more than type checking performance. By the third run, the source code and directory structures tend to be cached by the OS's file system cache, so that variability is eliminated.

Are your code bases untyped or typed (i.e. most functions have annotations for input parameters)? Are you using "strict" mode typing? How about the libraries that you're importing? Are they fully typed (i.e. have "py.typed" marker files or stubs)?

It would be interesting to see if there are specific files in your code base that are causing the majority of the variability. If you use the `--stats` and `--verbose` command-line options together, pyright will output a table at the end of its analysis that lists all of the files it checked sorted by the amount of time it took for each.

**Comment by excubo-jg** [comment]
Well, if these results deviate from your experience, it is quite likely that something is off at my end... To answer your questions:

The measurements were taken on a local windows machine, each repo in its own conda env. The libraries used do not overlap much, both are on python 3.10.9 and numpy 1.23.3. Node version did not change. 

What surprises me is that both repos exhibit behavior deviating from the norm as the two imo are very different themselves. One is a classic python cli app which is quite bulky as the subject data is part of the code and not read from disk and the second one is a flask web app... As the repos cannot be made public I have done some further analysis and provide more information using the flask app. If this is helpful I could do some more analysis. The flask repo is not typed, the other one is mostly. Type checking is set to basic, stubs are not used.

I rerun my analysis using the three-times approach you are describing. Times on average went up by 5% with individual scores fluctuating mostly by 10% or less and 25% at max. These changes had no effect on the form of the curve, neither did a run with the -stats parameter change the general picture (run time in sec):
![image](https://github.com/microsoft/pyright/assets/52322342/ae012775-e3a1-48f8-ad53-39d0aefcdd8d)

v293, 312, and 314 all have had an update to the type stubs so I am wondering if the fluctuations are driven by changes in the processing of libraries and not so much by the user code. 

The graph shows spikes in processing time with v282 and v287 - surprisingly only these version are impacted and times jump right back to the level seen before with the next version. The other app shows no unusual behavior with these versions.

I am attaching an xls-Sheet with the stats results. There are 10 files in the repo. The main flask file with the @app.route functionality has 350 LOC, total LOC is 1150.
[PyrightTest.xlsx](https://github.com/microsoft/pyright/files/12268910/PyrightTest.xlsx)

I am also attaching file #2. This has a processing time of 800ms up to 6300ms with significant jumps up and down between versions so hopefully you can infer some causality from it. The file imports file #3, none of the files imports the flask file #1. 
[File2.zip](https://github.com/microsoft/pyright/files/12269029/File2.zip)

Due to an oversight of mine the python-file to do the benchmarking with got included in the analysis... Its results are in the last line in the excel. The reported processing times go up by a factor of ten for this miniscule program, tripling with v306, This is the file: 
[PyrightPerf.zip](https://github.com/microsoft/pyright/files/12268964/PyrightPerf.zip)

**Reply by erictraut** [reply]
Thanks for the additional analysis.

I'm not able to reproduce most of the results you're seeing. I don't have access to your two repos, so I won't be able to attempt to repro those results. (The file `file2.py` references other missing files, so there's not much I can do with it.)

I focused on the `PyrightPerf` script that you provided since it's a complete code sample. I ran pyright versions 1.1.271 through 1.1.320 using the `--stats` command-line option. I ran each version five times and took the lowest time (which was typically the fifth run, but not always). The times were very consistent. From version 1.1.271 through 1.1.313, the times were between 487ms to 505ms, generally increasing as the version increased. That represents a 3.5% performance decrease over 50 versions. I'm not surprised to see a small performance loss over this long period of time given the number of features and capabilities I added during that time period.

I measured an unexpected performance drop between 1.1.313 and 1.1.314. The time went from 505ms to 544ms. That's a 7.2% performance regression in a single version — definitely something that requires additional investigation. I further isolated the regression to a [single commit](https://github.com/microsoft/pyright/commit/51d2ff8beb19b60e21732a11e75bb5c6514e62ac). @rchiodo, you'll be interested in this since this commit was copied from a PR you created. The commit involves a switch from ts-loader to esbuild-loader in our build process. This change noticeably sped up our builds, but it appears that esbuild-loader isn't doing as good a job transpiling or bundling the js code. If I revert this change, we regain all of the performance loss seen between 1.1.313 and 1.1.314.

I compared the ts-loader vs esbuild-loader performance on other source bases and found a similar performance difference (between 7 and 9% slower when using esbuild-loader). The esbuild-loader noticeably improves our build times, but it's not worth such a large runtime perf degradation. I'm therefore going to revert the esbuild-loader changes and go back to ts-loader for now. @rchiodo, you might want to do the same for pylance. Let me know if you have any ideas for how we can get back this performance while sticking with esbuild-loader.

@excubo-jg, I can't explain why you're seeing much larger perf variation between pyright versions than I'm seeing. I'm running on a Mac with M2 silicon, which is admittedly different from your setup. Maybe these variations are specific to Windows or Intel processors? That theory doesn't sound very plausible, but I'm grasping at straws. I ran my tests using node v20.5.0, which is the latest version, but as long as you're using a relatively new version of node (v16.x or newer), I wouldn't expect that to be an issue.

Your results show that pyright 1.1.287 and 1.1.282 were massive outliers. I wasn't able to repro this, but I looked at the change lists for versions 1.1.287 and 1.1.282. Version 1.1.287 introduced a pretty significant performance _improvement_, so if anything I'd expect analysis times to decrease. By comparison 1.1.288 didn't introduce any changes that should have any impact on performance. I don't see any changes in 1.1.282 or 1.1.283 that would point to a performance change. I don't have any working theories about these outliers you're seeing in your data. Without a way to repro the problem, I don't think there's much more I can do. Let me know if you have any further theories that you'd like me to explore.

**Comment by excubo-jg** [comment]
Many thanks for investigating this in more depth. 

I did rerun your analysis, also looking at the fastet of 5 runs. On my windows machine results were almost as stable as what you are reporting, with an 6% run-time increase until v313. v314 and thereafter jump to a 20% increase compared to v27x so the behavior is the same with Windows performance degrading a bit faster and the change you have implemented should bring times down again.

Apologies for sending a file which was not useful. I will try to find a shareable solution which reproduces the erratic behavior.

If I understand https://github.com/microsoft/pyright/discussions/5496 correctly, @has2k1 is reporting quite substantial changes in the processing time of public repos not only with v314/317 so I am wondering if you are covering those in your tests.

**Reply by erictraut** [reply]
Yes, version 1.1.317 included a performance improvement that focused on complex recursive protocol definitions. Recent versions of numpy have made increasing use of such protocols, and this led to a degradation in type checking perf for projects that make extensive use of numpy. The pylance team maintains a perf benchmark suite that includes projects dependent on numpy. They observed similar perf improvements for pyright 1.1.317.

Of the four public repos that were analyzed in #5496, only one (pandas) is currently covered in mypy-primer. The other three (mizani, plotnine, and py-shiny) are not. These could be good candidates for adding to mypy-primer for detecting behavior regressions, but I don't think they'd be good tests for perf regressions because they're too small. The entirety of the plotnine source base takes just 5sec to type check, so even small noise will result in seemingly large perf variance.

**Comment by excubo-jg** [comment]
I am going to send you an invite to a private github-repo which is derived from the second repo. This code sample tracks the progression of the processing times of the full repo on my machine and reproduces the spikes at v282, v287 and all other jumps in performance. The change in runtime between versions is 20% on average.

**Reply by erictraut** [reply]
Thanks. I'm able to repro the perf results with this code. Here's [the commit](https://github.com/microsoft/pyright/commit/cec5d7d5cb7e7b9c97ec5a5dcda7cd62c0995254) in 1.1.282 that caused the significant change in perf. You'll see in the commit comment that I speculated that it could potentially have an impact on analysis performance. I was able to mitigate that in 1.1.283 with a change.

I didn't dig any deeper into 1.1.287 because it's likely a similar issue, and these are very old versions of pyright, so I don't see value in investigating further.

Is there something specific that you wanted me to look at here with respect to the latest version of pyright (v1.1.321)? On my machine, the perf of v.1.1.321 is about 8% _faster_ than v1.1.281 when analyzing the code in your private repo (2.44s vs 2.23s).

**Reply by excubo-jg** [reply]
Many thanks for checking - good to know that things are in sync and performance differences are driven by changes in pyright and not by other factors. No need to revisit old versions, I was primarily reacting to your initial comment:

> The analysis times you've measured across pyright versions vary way more than any of the benchmarks we run. I wonder if you're using coding patterns or types that are not well represented in other benchmark or test suites. 

As I cannot share this repo publicly I thought it might be interesting for you to look for coding patterns et al.

**Comment by excubo-jg** [comment]
Even with v322 the first repo shows a significant increase in total processing time compared with earlier versions. Turns out the file which now takes the longest time to process until version v292 reported a processing time of about 5-6ms... Maybe a sanity check could capture instances when 2k lines of python code are processed instantaneously? I think that when initiated from the IDE pyright has behaved as expected, e.g. errors and warnings during coding have been reported.

If I exclude this file from the repo the improvement in performance is about 10% and in line with what you are seeing in the other repo.

**Comment by jbalogh** [comment]
Between 1.1.303 and 1.1.332 pyright runtime on my machine (Mac M1 Pro) almost doubled. I noticed this today while attempting to upgrade.

Here is an extremely unscientific benchmark of runtimes on my machine this morning via the following script. I didn't change the python source or virtualenv while it ran. I did four throw-away runs before recording the results of the fifth run. There were background processes like Firefox/Slack/Spotify running but nothing too heavy.

```python
for n in range(303, 333):
    print(n)
    os.system(f'npm install -g pyright@1.1.{n}')
    for _ in range(4):
        os.system("pyright app scripts lib 2>/dev/null 1>/dev/null")
    os.system(f"pyright  app scripts lib --stats 2>&1 | grep '^[A-Za-z0-9]' > pyright-{n}.txt")

# $ cat pyright* | awk '/pyright/{printf("| %s | ", $0)}/Completed/' | sed -e 's/Completed in //' -e 's/pyright//' -e 's/$/ |/' | pbcopy
# $ uname -a
# Darwin MacBook-Pro-3.local 22.5.0 Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:20 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6000 arm64
```

| version | time |
| ------- | ---- |
|  1.1.303 | 13.279sec |
|  1.1.304 | 13.187sec |
|  1.1.305 | 13.096sec |
|  1.1.306 | 16.206sec |
|  1.1.307 | 16.349sec |
|  1.1.308 | 17.834sec |
|  1.1.309 | 18.325sec |
|  1.1.310 | 17.783sec |
|  1.1.311 | 18.15sec |
|  1.1.312 | 21.892sec |
|  1.1.313 | 21.263sec |
|  1.1.314 | 24.578sec |
|  1.1.315 | 23.641sec |
|  1.1.316 | 23.056sec |
|  1.1.317 | 18.202sec |
|  1.1.318 | 18.283sec |
|  1.1.319 | 18.566sec |
|  1.1.320 | 17.936sec |
|  1.1.321 | 16.415sec |
|  1.1.322 | 16.11sec |
|  1.1.323 | 16.123sec |
|  1.1.324 | 23.719sec |
|  1.1.325 | 23.668sec |
|  1.1.326 | 23.669sec |
|  1.1.327 | 24.478sec |
|  1.1.328 | 24.771sec |
|  1.1.329 | 24.939sec |
|  1.1.330 | 24.704sec |
|  1.1.331 | 24.267sec |
|  1.1.332 | 24.103sec |


<details>
<summary>full stats for pyright 1.1.303</summary>

```
pyright 1.1.303
9 errors, 0 warnings, 0 informations 
Completed in 13.279sec
Analysis stats
Total files parsed and bound: 2021
Total files checked: 621
Timing stats
Find Source Files:    0.02sec
Read Source Files:    0.13sec
Tokenize:             0.36sec
Parse:                0.58sec
Resolve Imports:      0.71sec
Bind:                 0.6sec
Check:                10.51sec
Detect Cycles:        0sec

```

</details>

<details>
<summary>full stats for pyright 1.1.332</summary>

```
pyright 1.1.332
151 errors, 0 warnings, 0 informations 
Completed in 24.103sec
Analysis stats
Total files parsed and bound: 2920
Total files checked: 621
Timing stats
Find Source Files:    0.02sec
Read Source Files:    0.17sec
Tokenize:             0.79sec
Parse:                0.95sec
Resolve Imports:      0.92sec
Bind:                 1.25sec
Check:                19.86sec
Detect Cycles:        0sec

```
</details>

**Code or Screenshots**
With apologies, the code is in a private repo and I don't have a minimal repro. :\

This started out as an issue til I found this thread. Let me know if you'd rather me start a new thread or issue. 
We are big fans of pyright overall! Our only wish is that it would go faster.

**Reply by erictraut** [reply]
@jbalogh, thanks for reporting the issue. Unfortunately, without a repro there's not much I can do to diagnose the cause. If you're able to provide me with temporary private access to your repro, I'd be happy to dig into it. If that's not something you can do, I understand.

We have performance regression tests in place for pyright, and they don't show similar slowdowns to what you're reporting here. It's likely related to something that is specific to your code base. You could try running with `--stats ---verbose`. This will emit a lot of noise, but at the end of the analysis, it outputs a list of all files that it checked and the time it took for each (sorted greatest to least). You could do a comparison between two pyright versions and see if a particular source file pops to the top of the list. If so, you could then try to narrow it down further to a particular class or piece of code by commenting out sections of that module.

It looks like the biggest recent performance decrease is between 1.1.323 and 1.1.324. There were many changes between these two versions. I've reviewed all of the commits, and none of them jump out at me as likely candidates for a perf regression, but sometimes changes have non-obvious side effects. Type checkers are complex.

**Reply by erictraut** [reply]
@jbalogh, thanks for the code sample you sent me. This allowed me to find and fix the performance issue that you're seeing. Your code defined very large TypedDict classes (178 items). For each item defined in a TypedDict class, pyright synthesizes function overloads (e.g. for the `get` method). That means these classes had 178 overloads for some functions.

Another part of pyright validates that overloaded methods that override an overload in a base class are type compatible. This algorithm is O(m x n) where `m` is the number of overloads in the base class and `n` is the number of overloads in the subclass. There was a bug in this algorithm that I fixed in pyright 1.1.324 that resulted in some additional work. This additional work was small on its own, but when repeated `m x n` times, it added up!

I've changed the code to skip validation of synthesized overloads within a TypedDict class. Since these are synthesized by pyright, they're known to be type compatible with the `typing._TypedDict` base class, so any work done to validate them is wasted.

With this change, the type checking time for your sample file drops from 5080ms to 6ms on my machine. This change will be included in the next release.
